soc/arm/nuvoton_npcx/common/pinctrl_soc.h

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2022 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _NUVOTON_PINCTRL_SOC_H_
+#define _NUVOTON_PINCTRL_SOC_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief NPCX peripheral device configuration structure
+ *
+ * Used to indicate the peripheral device's corresponding register/bit for
+ * pin-muxing, pull-up/down and so on.
+ */
+struct npcx_periph {
+	/** Related register group for peripheral device. */
+	uint16_t group: 8;
+	/** Related register bit for peripheral device. */
+	uint16_t bit: 3;
+	/** The polarity for peripheral device functionality. */
+	uint16_t inverted: 1;
+	/** Reserved field. */
+	uint16_t reserved: 4;
+};
+
+/**
+ * @brief NPCX io-pad configuration structure
+ *
+ * Used to indicate a IO's corresponding GPIO module/bit for the pad properties
+ * such as bias-pull-up, drive-open-drain, low-volatge-input and so on.
+ */
+struct npcx_io_pad {
+	/** Indicate a IO pad's GPIO module */
+	uint16_t port: 5;
+	/** Indicate a IO pad's bit in GPIO module */
+	uint16_t bit: 3;
+	/** Related register index for low-voltage conf. */
+	uint16_t lvol_ctrl: 5;
+	/** Related register bit for low-voltage conf. */
+	uint16_t lvol_bit: 3;
+};
+
+/** Type for NPCX pin. */
+struct npcx_pinctrl {
+	union {
+		struct npcx_periph periph;
+		struct npcx_io_pad io_pad;
+		uint16_t cfg_word;
+	} cfg;
+	struct {
+		/** Indicates the current pinctrl type. */
+		uint16_t type : 4;
+		/** Properties used for pinmuxing. */
+		uint16_t pinmux_lock :1;
+		/** Properties used for io-pad. */
+		uint16_t io_type :2;
+		uint16_t io_bias_type :2;
+		uint16_t io_drive_type :1;
+		uint16_t io_1p8v_input :1;
+	} flags;
+};
+
+typedef struct npcx_pinctrl pinctrl_soc_pin_t;
+
+/** Types for NPCX pinctrl configurations. */
+#define NPCX_PINCTRL_TYPE_IO_PAD	0
+#define NPCX_PINCTRL_TYPE_PERIPH_PINMUX 1
+#define NPCX_PINCTRL_TYPE_PERIPH_PUPD	2
```

Comment:
Good idea. These flags will be moved to peripheral device structures. But I still reserve ''flag.type''' as two bits since the PSL configuration which structure is different from io-pad and peripheral devices will be introduced later. (Please refer to the draft PR [here](https://github.com/MulinChao/zephyr/tree/npcx_psl_gpio))

Reply:
Done.